### PR TITLE
Fix distro fact handling for Flatcar

### LINKED ISSUE #77537: Wrong OS family fact reported for Flatcar
### Summary

Ansible may report the wrong `ansible_os_family` fact on Flatcar Container Linux. The correct value is `Flatcar`, however some users are seeing `Flatcar Container Linux by Kinvolk` for the same fact under certain circumstances, which leads to wrong playbook behavior given that Ansible fails to identify Flatcar as the running distro.

In https://github.com/ansible/ansible/pull/69627 we've contributed Ansible core logic which affects the value of `ansible_os_family` for Flatcar. Back then we've specified `/etc/flatcar/update.conf` as the file based on which to figure out the name of the distro (likely because that's how CoreOS - Flatcar's direct ancestor - used to do the same). **This decision no longer makes sense to us** given that this file isn't the authoritative place for the distro name. Moreover, `/etc/flatcar/update.conf` may sometimes be missing based on user-provided bootstrap configuration.

### Issue Type

Bug Report

### Component Name

Ansible core

### Ansible Version

```console
$ ansible --version
ansible [core 2.11.5]
  config file = None
  configured module search path = ['/home/me/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/me/.local/lib/python3.8/site-packages/ansible
  ansible collection location = /home/me/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/me/.local/bin/ansible
  python version = 3.8.10 (default, Nov 26 2021, 20:14:08) [GCC 9.3.0]
  jinja version = 3.1.1
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed
```

### OS / Environment

Flatcar Container Linux 3139.2.0 for example

### Steps to Reproduce

Follow the official [instructions](https://www.flatcar.org/docs/latest/installing/vms/qemu/) to run Flatcar locally on QEMU:

```
mkdir flatcar; cd flatcar
wget https://stable.release.flatcar-linux.net/amd64-usr/3139.2.0/flatcar_production_qemu.sh
wget https://stable.release.flatcar-linux.net/amd64-usr/3139.2.0/flatcar_production_qemu.sh.sig
wget https://stable.release.flatcar-linux.net/amd64-usr/3139.2.0/flatcar_production_qemu_image.img.bz2
wget https://stable.release.flatcar-linux.net/amd64-usr/3139.2.0/flatcar_production_qemu_image.img.bz2.sig
bzip2 -d flatcar_production_qemu_image.img.bz2
chmod +x flatcar_production_qemu.sh

ssh-keygen -f key -q -N ""

./flatcar_production_qemu.sh -a ./key.pub -- -nographic
```

On another shell, SSH into the Flatcar VM:

```
cd flatcar
ssh -p 2222 -i key core@localhost
```

Install PyPy (Flatcar doesn't ship with a Python interpreter):

```
cd /opt
wget -O - https://downloads.python.org/pypy/pypy3.8-v7.3.9-linux64.tar.bz2 | sudo tar xjf -
sudo ln -s /opt/pypy3.8-v7.3.9-linux64/bin/pypy /opt/bin/python
```

Run Ansible against the VM and print the OS family:

```
cat <<EOF >playbook.yaml
- hosts: all
  user: core
  tasks:
  - name: Print distro
    debug:
      var: ansible_os_family
EOF

ansible-playbook playbook.yaml -i localhost:2222, --key-file ./key -e ansible_python_interpreter=/opt/bin/python -e ansible_port=2222
```

Output:

```
PLAY [all] *****************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************************************************************************************************************ok: [localhost]

TASK [Print distro] ********************************************************************************************************************************************************************************************************************************************ok: [localhost] => {
    "ansible_os_family": "Flatcar Container Linux by Kinvolk"
}

PLAY RECAP *****************************************************************************************************************************************************************************************************************************************************localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

SSH into the VM and populate `/etc/flatcar/update.conf` with dummy values:

```
printf "SERVER=foo\nGROUP=bar\n" | sudo tee /etc/flatcar/update.conf
```

Run the play book again:

```
ansible-playbook playbook.yaml -i localhost:2222, --key-file ./key -e ansible_python_interpreter=/opt/bin/python -e ansible_port=2222
```

Output:

```
PLAY [all] *****************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************************************************************************************************************ok: [localhost]

TASK [Print distro] ********************************************************************************************************************************************************************************************************************************************ok: [localhost] => {
    "ansible_os_family": "Flatcar"
}

PLAY RECAP *****************************************************************************************************************************************************************************************************************************************************localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### Expected Results

I expected to always see `Flatcar` as `ansible_os_family` on Flatcar Container Linux.

### Actual Results

```console
The value of `ansible_os_family` changes based on the existence of `/etc/flatcar/update.conf`.
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by johananl**
##### SUMMARY
The existence of the file `/etc/flatcar/update.conf` depends on bootstrap configuration typically provided by the user. For that reason this file is unsuitable for determining distro facts for Flatcar Container Linux. This PR switches to `/etc/os-release` as the file used for extracting distro facts for Flatcar.

Fixes #77537.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

distribution.py

>NOTE: I'm having hard time figuring out how to test this change. There is `test/units/module_utils/facts/system/distribution/fixtures/flatcar_2492.0.0.json` but I couldn't get reliable results by executing the test which uses this fixture. Any help regarding how to test this change would be appreciated!

### CODE REVIEW

File: lib/ansible/module_utils/facts/system/distribution.py [hunk 1]
```diff
@@ -447,15 +447,16 @@ def parse_distribution_file_Flatcar(self, name, data, path, collected_facts):
...
-            if release:
-                flatcar_facts['distribution_release'] = release.group(1).strip('"')
-        else:
+        if distro.lower() != 'flatcar':
             return False, flatcar_facts
 
+        if not data:
+            return False, flatcar_facts
+
+        version = re.search("VERSION=(.*)", data)
```
Comment (pothos) [inline lib/ansible/module_utils/facts/system/distribution.py:456]: Maybe we could even rely more on `distro` and reuse `_distro.version()` that already provides the `VERSION_ID` field, or?

Comment (johananl) [reply lib/ansible/module_utils/facts/system/distribution.py:456]: I like the idea, however `distro` inside `parse_distribution_file_Flatcar()` is a **string**, not an instance of `LinuxDistribution`. The `version()` method on `LinuxDistribution` is actually already used inside `get_distribution_version()` which gets invoked from `_guess_distribution()` which in turns runs before we try processing the dist file. The dist file parsing functions are **overriding** any facts guessed previously, and AFAICT are supposed to be the authoritative source for distro-specific facts.

Therefore, I think it's safer to explicitly set the relevant facts inside `parse_distribution_file_Flatcar()` and rely on guessed facts only as a fallback.

**Comment by johananl** [comment]
>>NOTE: I'm having hard time figuring out how to test this change. There is `test/units/module_utils/facts/system/distribution/fixtures/flatcar_2492.0.0.json` but I couldn't get reliable results by executing the test which uses this fixture. Any help regarding how to test this change would be appreciated!

I've updated the existing test fixture. The test seems to pass, however I'm not totally sure about the `distro` key [here](https://github.com/ansible/ansible/pull/77635/files#diff-6821b161be4d93232945e9a11ac642ca6afeaa70c07690341df67ef82c27c073R7) and whether it is correct.

**Comment by johananl** [comment]
> > The test seems to pass, however I'm not totally sure about the `distro` key [here](https://github.com/ansible/ansible/pull/77635/files#diff-6821b161be4d93232945e9a11ac642ca6afeaa70c07690341df67ef82c27c073R7) and whether it is correct.
> 
> That fixture can be generated by running `hacking/tests/gen_distribution_version_testcase.py`

Thanks @sivel. The generated output has `result.distribution` and `result.os_family` set to `Flatcar Container Linux by Kinvolk` though, which isn't what we want (see #77537) 😕
Generating the fixture and then setting these two keys to `Flatcar` manually seems to work well though.

**Pull request merged**